Euify and Reactify Query Bar Component

Bargs opened the pull request:
Implements query bar portion of https://elastic.github.io/eui/#/layout/header. Filter bar will come in another PR.

Fixes https://github.com/elastic/kibana/issues/14086

Re-implements our query bar component in React using some EUI components. Existing typeahead and suggestion styles were copied over 1:1 for now after talking with Dave about it. In this PR I focused on reaching feature parity with the existing query bar. Some additional work would be needed before we could move this into EUI as a generic component that could be consumed by other plugins.

Still needs some new tests and I suspect some old tests will need to be updated, but other than that this PR is functionally complete and ready for reviews.

trevan commented:
Could you "virtualize" the suggestions with something like react-virtualized?

Bargs commented:
@trevan was already working on it ;) https://github.com/elastic/kibana/pull/23704/commits/3e8ad02e164157bfe70e365605e47943bf89e35d

Bargs commented:
BTW I just tested this on my large fields test case and it resolved the issues we discussed in https://github.com/elastic/kibana/issues/14086

Let me know if you still see any issues.

cjcenizal commented on src/ui/public/query_bar/components/typeahead/suggestion.js:
```diff
@@ -0,0 +1,81 @@
+  }
+}
+
+export function Suggestion(props) {
```
Could I make a suggestion? (Pardon the pun). Here's how I've seen a lot of components formatted both in EUI and in Kibana:

```jsx
export function Suggestion({
  selected,
  suggestion,
  onMouseEnter,
  onClick,
  innerRef,
  ariaId,
}) {
  const {
    text,
    description,
    type,
  } = suggestion;

  const classes = classNames('typeahead-item', { active: selected });
  const itemClasses = `suggestionItem suggestionItem--${type}`;

  return (
    <div
      className={classes}
      role="option"
      onClick={() => onClick(suggestion)}
      onMouseEnter={onMouseEnter}
      ref={innerRef}
      id={ariaId}
    >
      <div className={itemClasses}>
        <div className="suggestionItem__type" type={type}>
          <EuiIcon type={getEuiIconType(type)}/>
        </div>
        <div className="suggestionItem__text">{text}</div>
        <div
          className="suggestionItem__description"
          // Description currently always comes from us and we escape any potential user input
          // at the time we're generating the description text
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{ __html: description }}
        />
      </div>
    </div>
  );
}
```

The notable differences are the use of destructuring and the extraction of calculated values out of the JSX itself. At first I didn't like the use of destructuring but it grew on me because it helped me identify which props I was using up-front, and it also reduced noise in the JSX. Just a suggestion, feel free to disregard.

Bargs replied:
I'm mostly cool with destructuring and happy to follow whatever conventions folks are using for react stuff already, will make the update 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3944/)

snide commented:
Hey @Bargs nice work. Want me to pass you a PR to convert the Less to Sass? Should be a pretty quick, non-breaking switch.

Bargs commented:
@snide for sure!

timroes commented on src/ui/public/query_bar/components/typeahead/suggestion.js:
```diff
@@ -0,0 +1,81 @@
+          // Description currently always comes from us and we escape any potential user input
+          // at the time we're generating the description text
+          // eslint-disable-next-line react/no-danger
+          dangerouslySetInnerHTML={{ __html: props.suggestion.description }}
```
Out of interest: Where are those descriptions actually set/coming from? Could we potentially write descriptions in JSX directly, so we wouldn't need to use `dangerouslySetInnerHTML` here?

Bargs replied:
They come from the autocomplete provider types https://github.com/elastic/kibana/blob/ecc0f5e6e4bc49023cf67bf808690d434068a6be/x-pack/plugins/kuery_autocomplete/public/autocomplete_providers/field.js#L16

timroes commented:
Accessibility looks fine for me. `aria-activedescendant` is set to a value `suggestion-null` even if no element is selected, but since that is not a valid id, a screenreader won't break on that. Of course it would be nicer not to set it at all if no autocompletion is selected, but that's a minor nitpick (and most likely not even related to this PR).

I notice two other kind of strange behaviors, but both of them might actually also be EUI issues:

1. If you focus the query bar, than press <kbd>Tab</kbd> the Options button is focused (so far so good). If you know press <kbd>Enter</kbd> I would expect the popover to open, but instead the search is executed, even though the Options button was focused.

2. If you open the popover, and then click into the Query Bar, the query bar is only shortly focused, before losing its focus again.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4239/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4742/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4930/)

Bargs commented:
All done typescriptifying and writing tests for the new code, just need to get CI green now.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5389/)

Bargs commented:
jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5613/)

lukasolson commented on src/core_plugins/kibana/public/visualize/editor/editor.html:
```diff
@@ -40,11 +40,10 @@
-            index-patterns="[indexPattern]"
-          >
-          </query-bar>
+            index-patterns="[indexPatter]"
```
Should be `indexPattern`

lukasolson commented on src/ui/public/query_bar/components/query_bar.tsx:
```diff
@@ -0,0 +1,490 @@
+
+import { EuiFieldText, EuiOutsideClickDetector } from '@elastic/eui';
+
+const KEY_CODES = {
```
Pretty sure these can be imported from EUI: `import { keyCodes } from '@elastic/eui';`

Bargs replied:
Just checked, unfortunately EUI doesn't include `home` or `end`

lukasolson commented:
Playing around with it, it looks like we're not resetting the selected suggestion when selecting a suggestion (that sounds weird, but I hope it makes sense). We used to do this and I think it makes sense so that the user doesn't have to hit <kbd>esc</kbd> then <kbd>enter</kbd> to submit the query. Thoughts?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5625/)

Bargs commented:
@lukasolson Addressed all your feedback

> it looks like we're not resetting the selected suggestion when selecting a suggestion

Good catch, I knew something felt off but I couldn't figure out what it was. 
https://github.com/elastic/kibana/pull/23704/commits/22038d56b99e9668ab91fbbd005a4cd26306a375

> Also I noticed that there's no longer a search icon to the right of the query bar. I'm assuming this is by intention.

Yeah this is following the new design from @cchaos : https://elastic.github.io/eui/#/layout/header

> I've also noticed that the bug still exists on Dashboard that if I hit enter to refresh the same query it's not actually submitting it. 

Do you know if this is something that needs fixed in the query bar? If it works in discover and visualize but not dashboard, I would suspect it's an issue in the dashboard code.

lukasolson commented:
> Do you know if this is something that needs fixed in the query bar? If it works in discover and visualize but not dashboard, I would suspect it's an issue in the dashboard code.

I think you're probably right, we can address this separately.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5652/)

Bargs commented:
Looks like the failure is a [flaky test](https://github.com/elastic/kibana/issues/24292), going to retry if github will let me right now.

timroes commented on src/ui/public/query_bar/components/typeahead/suggestion_component.tsx:
```diff
@@ -0,0 +1,80 @@
+          // at the time we're generating the description text
+          // eslint-disable-next-line react/no-danger
+          // @ts-ignore
+          dangerouslySetInnerHTML={{ __html: props.suggestion.description }}
```
Since this exported from the actual suggestions as @Bargs showed, couldn't we just change those suggestions classes to export a JSX element instead of a string?

```js
getDescription() {
  return (<React.Fragment>Some description that <b>could contain</b> HTML.</React.Fragment>);
}
```

That way we would not need to use `dangerouslySetInnerHTML` here and could just use that description directly. I think every place we could get rid of it, would be one less place that could expose a vulnerability in the end :-)

Bargs replied:
We definitely could, but then I think we would need to update every single consumer of these suggestions. That includes APM, beats_management, and infra, in addition to this query bar component. I'm hesitant to make that many changes to code I'm not familiar with this late in the game. Since this issue predates this PR (it was "dangerously set" in angular as well) I think it would be ok to handle it in a separate PR, do you agree?

timroes replied:
Yeah totally fine for me doing this in a separate PR and and making other teams aware.

Bargs commented:
jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5824/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5900/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5994/)

Stacey-Gammon commented:
Looks good.  I have one comment that doesn't have  to go in now, but that I find a bit annoying.  I type in a new query and hit enter and the next thing that happens is the auto complete box pops up:

<img width="245" alt="screen shot 2018-10-23 at 9 40 17 am" src="https://user-images.githubusercontent.com/16563603/47364569-aae7ca80-d6a7-11e8-8bbe-ec05aa27e6bf.png">

Almost seems like hitting enter should close that box and not re-open it until the user starts typing again.  Otherwise I just executed that query, why do I need to see that box, and it kinda hides the filter bar.  

One other minor comment that I find annoying.  After I select the toggle to turn on Kuery, I then click into the query bar, but the first thing that is highlighted is the Options button, not the query bar itself, so it requires two clicks to start typing.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6038/)

Bargs commented:
Thanks for the reviews everyone! I'm going to make a list of the non-critical suggestions to follow up on in another PR.

trevan commented:
@Bargs, @cchaos 
> > Also I noticed that there's no longer a search icon to the right of the query bar. I'm assuming this is by intention.
> 
> Yeah this is following the new design from @cchaos : https://elastic.github.io/eui/#/layout/header

I tried out the changes and I think the missing search icon is a problem.  I understand it is the new design but I think it should be changed.  To "refresh" the data, you could just click the search icon.  Now, I have to click into the search bar and then hit enter.  There needs to be some sort of icon on the bar that reruns the query.

trevan commented:
@cchaos, that works and it really is needed before this hits release.

trevan commented on src/ui/public/query_bar/components/query_bar.tsx:
```diff
@@ -0,0 +1,502 @@
+      return recentQueryString.includes(query);
+    });
+    return matchingRecentSearches.map(recentSearch => {
+      const text = recentSearch;
```
@Bargs, you should also call toUser() on recentSearch if it is an object.  Otherwise, react doesn't like it when the object is used in the Suggestion component.

Bargs commented:
@Stacey-Gammon I was just going through some of the feedback and creating issues for the minor things I didn't have time for. However I couldn't reproduce this one issue you mentioned:

> I type in a new query and hit enter and the next thing that happens is the auto complete box pops up

Do you have any additional details about that? Browser you were using maybe? It's not happening for me in master in Safari at least.
